lxd/instance/drivers/driver_qemu.go

```diff
@@ -1763,6 +1773,102 @@ func (d *qemu) deviceStart(deviceName string, rawConfig deviceConfig.Device, ins
 	return runConf, nil
 }
 
+func (d *qemu) deviceAttachBlockDevice(deviceName string, configCopy map[string]string, mounts []deviceConfig.MountEntryItem) error {
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Check if the agent is running.
+	monitor, err := qmp.Connect(d.monitorPath(), qemuSerialChardevName, d.getMonitorEventHandler())
+	if err != nil {
+		return fmt.Errorf("Failed to connect to QMP monitor: %w", err)
+	}
+
+	fields := strings.Split(mounts[0].DevPath, ":")
+	blockFilename := fields[2]
+	oldDevPaths := d.devPaths
```

Comment:
I'm a bit concerned by this, as according to the comment on the `DevPaths()` function, d.devPaths is only fully populated with the device paths for all devices during instance startup. However presumably the hot plug feature will be called both at start up and after start up - at which point the d.devPaths will be empty, and you'll potentially end up replacing formerly allowed device paths added to the apparmor profile during start with just the hot plugged disk device path. Have I read this right?

Also a comment here explaining what is going on would be useful as it required me to read around the instance code base to figure out what was going on here.

Reply:
> However presumably the hot plug feature will be called both at start up and after start up - at which point the d.devPaths will be empty, and you'll potentially end up replacing formerly allowed device paths added to the apparmor profile during start with just the hot plugged disk device path. Have I read this right?

The hot plug feature is only called when a device is being hot-plugged, i.e. not at VM startup.

Reply:
Ideally we would have this feature behave like the NIC hot-plug feature, which looking at `addNetDevConfig()` (which is the function that then calls `m.AddNIC()` on the `qmp` package, its called twice, at start up and at hot-plug time.
This reason I went this route rather than adding device config to the qemu.conf file, is to minimise the variations in device management. Effectively making the QMP add/remove approach the *only* way we add/remove NICs.

This is also preferred by the upstream QEMU team who are looking at phasing out the qemu.conf altogether, and we've seen a few breakages in parsing that file. Whereas using QMP is the recommended way to manage devices.

So for this PR I would expect to be seeing the disk devices (or at least the ones that can be unplugged) being removed from static config file and switching to use QMP at start up to add them.

Reply:
> OK, I'll rework the code to use QMP for all disks except maybe the root disk.

Thanks, if you can use it for the root disk even better, as it avoids the variety of code paths we have to support.